configloader data race

Found this data race in CI: 

```bash
WARNING: DATA RACE
Read at 0x00c0002be988 by goroutine 24:
  github.com/envoyproxy/gateway/internal/envoygateway/config/loader.TestConfigLoaderStandaloneExtensionServerAndCustomResource.func3()
      /home/runner/work/gateway/gateway/internal/envoygateway/config/loader/configloader_test.go:95 +0x10c
  github.com/envoyproxy/gateway/internal/envoygateway/config/loader.(*Loader).runHook.func1()
      /home/runner/work/gateway/gateway/internal/envoygateway/config/loader/configloader.go:117 +0xa2
  github.com/envoyproxy/gateway/internal/envoygateway/config/loader.(*Loader).runHook.gowrap1()
      /home/runner/work/gateway/gateway/internal/envoygateway/config/loader/configloader.go:120 +0x4f

Previous write at 0x00c0002be988 by goroutine 21:
  reflect.Value.Set()
      /opt/hostedtoolcache/go/1.25.5/x64/src/reflect/value.go:2139 +0x1a8
  sigs.k8s.io/json/internal/golang/encoding/json.indirect()
      /home/runner/go/pkg/mod/sigs.k8s.io/json@v0.0.0-20250730193827-2d320260d730/internal/golang/encoding/json/decode.go:503 +0x448
  sigs.k8s.io/json/internal/golang/encoding/json.(*decodeState).object()
      /home/runner/go/pkg/mod/sigs.k8s.io/json@v0.0.0-20250730193827-2d320260d730/internal/golang/encoding/json/decode.go:636 +0x78
  sigs.k8s.io/json/internal/golang/encoding/json.(*decodeState).value()
      /home/runner/go/pkg/mod/sigs.k8s.io/json@v0.0.0-20250730193827-2d320260d730/internal/golang/encoding/json/decode.go:401 +0xa4
  sigs.k8s.io/json/internal/golang/encoding/json.(*decodeState).object()
      /home/runner/go/pkg/mod/sigs.k8s.io/json@v0.0.0-20250730193827-2d320260d730/internal/golang/encoding/json/decode.go:850 +0x1d38
  sigs.k8s.io/json/internal/golang/encoding/json.(*decodeState).value()
      /home/runner/go/pkg/mod/sigs.k8s.io/json@v0.0.0-20250730193827-2d320260d730/internal/golang/encoding/json/decode.go:401 +0xa4
  sigs.k8s.io/json/internal/golang/encoding/json.(*decodeState).unmarshal()
      /home/runner/go/pkg/mod/sigs.k8s.io/json@v0.0.0-20250730193827-2d320260d730/internal/golang/encoding/json/decode.go:184 +0x296
  sigs.k8s.io/json/internal/golang/encoding/json.Unmarshal()
      /home/runner/go/pkg/mod/sigs.k8s.io/json@v0.0.0-20250730193827-2d320260d730/internal/golang/encoding/json/decode.go:112 +0x2c4
  sigs.k8s.io/json.UnmarshalCaseSensitivePreserveInts()
      /home/runner/go/pkg/mod/sigs.k8s.io/json@v0.0.0-20250730193827-2d320260d730/json.go:62 +0x25a
  k8s.io/apimachinery/pkg/runtime/serializer/json.(*Serializer).unmarshal()
      /home/runner/go/pkg/mod/k8s.io/apimachinery@v0.34.2/pkg/runtime/serializer/json/json.go:270 +0x1ad
  k8s.io/apimachinery/pkg/runtime/serializer/json.(*Serializer).Decode()
      /home/runner/go/pkg/mod/k8s.io/apimachinery@v0.34.2/pkg/runtime/serializer/json/json.go:209 +0x99b
  k8s.io/apimachinery/pkg/runtime/serializer/recognizer.(*decoder).Decode()
      /home/runner/go/pkg/mod/k8s.io/apimachinery@v0.34.2/pkg/runtime/serializer/recognizer/recognizer.go:110 +0x2be
  github.com/envoyproxy/gateway/internal/envoygateway/config.DecodeBytes()
      /home/runner/work/gateway/gateway/internal/envoygateway/config/decoder.go:41 +0x16d
  github.com/envoyproxy/gateway/internal/envoygateway/config.Decode()
      /home/runner/work/gateway/gateway/internal/envoygateway/config/decoder.go:34 +0xf2
  github.com/envoyproxy/gateway/internal/envoygateway/config/loader.(*Loader).Start.func1()
      /home/runner/work/gateway/gateway/internal/envoygateway/config/loader/configloader.go:65 +0x5d8

Goroutine 24 (running) created at:
  github.com/envoyproxy/gateway/internal/envoygateway/config/loader.(*Loader).runHook()
      /home/runner/work/gateway/gateway/internal/envoygateway/config/loader/configloader.go:116 +0x264
  github.com/envoyproxy/gateway/internal/envoygateway/config/loader.(*Loader).Start.func1()
      /home/runner/work/gateway/gateway/internal/envoygateway/config/loader/configloader.go:93 +0xc84

Goroutine 21 (running) created at:
  github.com/envoyproxy/gateway/internal/envoygateway/config/loader.(*Loader).Start()
      /home/runner/work/gateway/gateway/internal/envoygateway/config/loader/configloader.go:55 +0x39c
  github.com/envoyproxy/gateway/internal/filewatcher.(*fileWatcher).Add()
      /home/runner/work/gateway/gateway/internal/filewatcher/filewatcher.go:87 +0x176
  github.com/envoyproxy/gateway/internal/envoygateway/config/loader.(*Loader).Start()
      /home/runner/work/gateway/gateway/internal/envoygateway/config/loader/configloader.go:50 +0x279
  github.com/envoyproxy/gateway/internal/envoygateway/config/loader.TestConfigLoaderStandaloneExtensionServerAndCustomResource()
      /home/runner/work/gateway/gateway/internal/envoygateway/config/loader/configloader_test.go:101 +0x484
  testing.tRunner()
      /opt/hostedtoolcache/go/1.25.5/x64/src/testing/testing.go:1934 +0x21c
  testing.(*T).Run.gowrap1()
      /opt/hostedtoolcache/go/1.25.5/x64/src/testing/testing.go:1997 +0x44
==================
--- FAIL: TestConfigLoaderStandaloneExtensionServerAndCustomResource (3.02s)
    configloader_test.go:93: config changed 1 times
    configloader_test.go:93: config changed 2 times
    testing.go:1617: race detected during execution of test
```

--- a/internal/envoygateway/config/loader/configloader.go
+++ b/internal/envoygateway/config/loader/configloader.go
@@ -8,6 +8,7 @@ package loader
 import (
 	"context"
 	"io"
+	"sync"
 	"time"
 
 	"github.com/envoyproxy/gateway/api/v1alpha1/validation"
@@ -24,6 +25,7 @@ type Loader struct {
 	logger  logging.Logger
 	cancel  context.CancelFunc
 	hook    HookFunc
+	mu      sync.RWMutex
 
 	w filewatcher.FileWatcher
 }
@@ -76,10 +78,13 @@ func (r *Loader) Start(ctx context.Context, logOut io.Writer) error {
 
 				// Set defaults for unset fields
 				eg.SetEnvoyGatewayDefaults()
+				eg.Logging.SetEnvoyGatewayLoggingDefaults()
+
+				r.mu.Lock()
 				r.cfg.EnvoyGateway = eg
 				// update cfg logger
-				eg.Logging.SetEnvoyGatewayLoggingDefaults()
 				r.cfg.Logger = logging.NewLogger(logOut, eg.Logging)
+				r.mu.Unlock()
 
 				// cancel last
 				if r.cancel != nil {
@@ -111,11 +116,28 @@ func (r *Loader) runHook(ctx context.Context) {
 	}
 
 	r.logger.Info("running hook")
+	cfgCopy := r.snapshotConfig()
 	c, cancel := context.WithCancel(ctx)
 	r.cancel = cancel
 	go func(ctx context.Context) {
-		if err := r.hook(ctx, r.cfg); err != nil {
+		if err := r.hook(ctx, cfgCopy); err != nil {
 			r.logger.Error(err, "hook error")
 		}
 	}(c)
+}
+
+func (r *Loader) snapshotConfig() *config.Server {
+	r.mu.RLock()
+	defer r.mu.RUnlock()
+
+	if r.cfg == nil {
+		return nil
+	}
+
+	cp := *r.cfg
+	if r.cfg.EnvoyGateway != nil {
+		cp.EnvoyGateway = r.cfg.EnvoyGateway.DeepCopy()
+	}
+
+	return &cp
 }

